pkg/server/plugin/noderesolver/aws/iid.go

```diff
@@ -0,0 +1,282 @@
+	}
+	if config.SecretAccessKey == "" {
+		config.SecretAccessKey = p.hooks.getenv("AWS_SECRET_ACCESS_KEY")
+	}
+
+	conf := aws.NewConfig()
+	switch {
+	case config.AccessKeyID != "" && config.SecretAccessKey != "":
+		conf.Credentials = credentials.NewStaticCredentials(config.AccessKeyID, config.SecretAccessKey, "")
+	case config.AccessKeyID != "" && config.SecretAccessKey == "":
+		return nil, iidError.New("configuration missing secret access key")
+	case config.AccessKeyID == "" && config.SecretAccessKey != "":
+		return nil, iidError.New("configuration missing access key id")
+	case config.AccessKeyID == "" && config.SecretAccessKey == "":
+		return nil, iidError.New("configuration missing both access key id and secret access key")
+	}
+
+	// set the AWS configuration and reset clients
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.conf = conf
+	p.clients = make(map[string]awsClient)
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) Resolve(ctx context.Context, req *noderesolver.ResolveRequest) (*noderesolver.ResolveResponse, error) {
+	resp := &noderesolver.ResolveResponse{
+		Map: make(map[string]*common.Selectors),
+	}
+	for _, spiffeID := range req.BaseSpiffeIdList {
+		selectors, err := p.resolveSpiffeID(ctx, spiffeID)
+		if err != nil {
+			return nil, err
+		}
+		if selectors != nil {
+			resp.Map[spiffeID] = selectors
+		}
+	}
+	return resp, nil
+}
+
+func (p *IIDResolverPlugin) resolveSpiffeID(ctx context.Context, spiffeID string) (*common.Selectors, error) {
+	_, region, instanceID, err := parseAgentID(spiffeID)
+	if err != nil {
+		logrus.Warnf("unrecognized Agent ID: %s: %v", spiffeID, err)
+		return nil, nil
+	}
+
+	client, err := p.getRegionClient(region)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(instanceID)},
+		Filters:     instanceFilters,
+	})
+	if err != nil {
+		return nil, iidError.Wrap(err)
+	}
+
+	selectorSet := map[string]bool{}
+	addSelectors := func(values []string) {
+		for _, value := range values {
+			selectorSet[value] = true
+		}
+	}
+
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			addSelectors(resolveTags(instance.Tags))
+			addSelectors(resolveSecurityGroups(instance.SecurityGroups))
+			if instance.IamInstanceProfile != nil && instance.IamInstanceProfile.Arn != nil {
+				output, err := client.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{
+					InstanceProfileName: instance.IamInstanceProfile.Arn,
+				})
+				if err != nil {
+					return nil, iidError.Wrap(err)
+				}
+				addSelectors(resolveInstanceProfile(output.InstanceProfile))
+			}
+		}
+	}
+
+	// sort and dedup selectors
```

Comment:
This comment is no longer true I think, because deduplication was already done because of the use of a map. Maybe we could just put "// sort selectors"

Reply:
good catch. will update.